src/stores/wordpress-versions-api.ts

```diff
@@ -0,0 +1,167 @@
+import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
+import * as Sentry from '@sentry/electron/renderer';
+import { __, sprintf } from '@wordpress/i18n';
+import { z } from 'zod';
+import { isWordPressDevVersion, isWordPressBetaVersion } from 'src/lib/wordpress-version-utils';
+import { withOfflineCheck } from 'src/stores/tests/utils/with-offline-check';
+
+const MINIMUM_WORDPRESS_VERSION = '5.9.9';
+
+const wordPressOfferSchema = z.object( {
+	version: z.string(),
+	response: z.string(),
+} );
+
+type WordPressOffer = z.infer< typeof wordPressOfferSchema >;
+
+const wordPressApiResponseSchema = z.object( {
+	offers: z.array( z.any() ),
+} );
+
+type WordPressApiOffer = {
+	version: string;
+	response?: string;
+	[ key: string ]: unknown;
+};
+
+type ProcessedOffer = {
+	version: string;
+	shortName: string;
+};
+
+const extractShortName = ( version: string ): string => {
+	if ( isWordPressDevVersion( version ) ) {
+		return 'nightly';
+	}
+	const match = version.match( /^(\d+\.\d+)/ );
+	return match ? match[ 1 ] : version;
+};
+
+export interface WordPressVersion {
+	isBeta: boolean;
+	isDevelopment: boolean;
+	isLatest: boolean;
+	label: string;
+	value: string;
+}
+
+async function fetchWordPressApiData( channel: 'beta' | 'development', version?: string ) {
+	const baseUrl = 'https://api.wordpress.org/core/version-check/1.7/';
+	const params = new URLSearchParams( { channel } );
+
+	if ( channel === 'beta' && version ) {
+		params.append( 'version', version );
+	}
+
+	const response = await fetch( `${ baseUrl }?${ params }` );
+	if ( ! response.ok ) {
+		throw new Error( 'Failed to fetch WordPress versions' );
+	}
+	return wordPressApiResponseSchema.parse( await response.json() );
+}
+
+function processWordPressOffers(
+	offers: WordPressApiOffer[],
+	isDevelopment = false,
+	shortNameOccurrences: Map< string, number >
+): ProcessedOffer[] {
+	return offers
+		.map( ( offer ) => wordPressOfferSchema.safeParse( offer ) )
+		.filter( ( result ): result is { success: true; data: WordPressOffer } => result.success )
+		.filter( ( result ) =>
+			isDevelopment ? result.data.response === 'development' : result.data.response === 'autoupdate'
+		)
+		.map( ( { data } ) => {
+			const shortName = extractShortName( data.version );
+			shortNameOccurrences.set( shortName, ( shortNameOccurrences.get( shortName ) || 0 ) + 1 );
+			return {
+				version: data.version,
+				shortName,
+			};
+		} );
+}
+
+function generateVersionLabel(
+	version: string,
+	shortName: string,
+	occurrences: number,
+	isLatest: boolean
+): string {
+	if ( isWordPressDevVersion( version ) ) {
+		return 'nightly';
+	}
+	if ( isLatest ) {
+		return sprintf( __( '%s (latest)' ), version );
+	}
+	if ( occurrences > 1 || isWordPressBetaVersion( version ) ) {
+		return version;
+	}
```

Comment:
This isn't exactly an issue, it's more that I haven't really understood the logic around `shortNameOccurences` and `occurrences`. Maybe it deserves a comment for clarification?

Reply:
Thanks for pointing it out. I've added some comments https://github.com/Automattic/studio/pull/1102/commits/78e3e6615b76812026a0df1d2df0219db42f4769
and renamed the parameter for consistency https://github.com/Automattic/studio/pull/1102/commits/f0efa981a196265b5d31d058e19bed5496a122b0

Let me know if it's not clear enough.

Reply:
Thanks, that clarifies the logic 👍